source/nodejs/adaptivecards/src/shared.ts

```diff
@@ -2,170 +2,183 @@
+    applicationJson: "application/json",
+    applicationXWwwFormUrlencoded: "application/x-www-form-urlencoded"
 }
 
 export interface ISeparationDefinition {
-	spacing: number,
-	lineThickness?: number,
-	lineColor?: string
+    spacing: number,
+    lineThickness?: number,
+    lineColor?: string
 }
 
 export interface IInput {
-	id: string;
-    value: string;
+    id?: string;
+    value?: string;
     validateValue(): boolean;
 }
 
 export type Dictionary<T> = { [key: string]: T };
 
 export class StringWithSubstitutions {
-	private _isProcessed: boolean = false;
-	private _original: string = null;
-    private _processed: string = null;
+    private _isProcessed: boolean = false;
+    private _original?: string;
+    private _processed?: string;
 
     getReferencedInputs(inputs: IInput[], referencedInputs: Dictionary<IInput>) {
         if (!referencedInputs) {
             throw new Error("The referencedInputs parameter cannot be null.")
         }
 
         for (let input of inputs) {
-            let matches = new RegExp("\\{{2}(" + input.id + ").value\\}{2}", "gi").exec(this._original);
+            let matches = new RegExp("\\{{2}(" + input.id + ").value\\}{2}", "gi").exec(<string>this._original);
 
-            if (matches != null) {
+            if (matches != null && input.id) {
                 referencedInputs[input.id] = input;
             }
         }
     }
 
-	substituteInputValues(inputs: Dictionary<IInput>, contentType: string) {
-		this._processed = this._original;
+    substituteInputValues(inputs: Dictionary<IInput>, contentType: string) {
+        this._processed = this._original;
+
+        if (this._original) {
+            let regEx = /\{{2}([a-z0-9_$@]+).value\}{2}/gi;
+            let matches;
+
+            while ((matches = regEx.exec(<string>this._original)) != null) {
+                let matchedInput: IInput | undefined = undefined;
+
+                for (let key of Object.keys(inputs)) {
+                    if (key.toLowerCase() == matches[1].toLowerCase()) {
+                        matchedInput = inputs[key];
+                        break;
+                    }
+                }
+
+                if (matchedInput) {
+                    let valueForReplace = "";
+
+                    if (matchedInput.value) {
+                        valueForReplace = matchedInput.value;
+                    }
+
+                    if (contentType === ContentTypes.applicationJson) {
+                        valueForReplace = JSON.stringify(valueForReplace);
+                        valueForReplace = valueForReplace.slice(1, -1);
+                    }
+                    else if (contentType === ContentTypes.applicationXWwwFormUrlencoded) {
+                        valueForReplace = encodeURIComponent(valueForReplace);
+                    }
+
+                    this._processed = (<string>this._processed).replace(matches[0], valueForReplace);
+                }
```

Comment:
nit: consider moving these lines to LN67

Reply:
Done: https://github.com/microsoft/AdaptiveCards/pull/3508/commits/431880c3068575ae57938a1a169857dcea238b86